src/backend/auxil/range_alloc.rs

```diff
@@ -0,0 +1,190 @@
+use std::fmt::Debug;
+use std::ops::{Add, AddAssign, Range, Sub};
+
+pub struct RangeAllocator<T> {
+    /// The range this allocator covers.
+    initial_range: Range<T>,
+    /// A Vec of ranges in this heap which are unused.
+    /// Must be ordered with ascending range start to permit short circuiting allocation.
+    /// No two ranges in this vec may overlap.
+    free_ranges: Vec<Range<T>>,
+}
+
+impl<T> RangeAllocator<T>
+where
+    T: Clone + Copy + Add<Output = T> + AddAssign + Sub<Output = T> + Eq + PartialOrd + Debug,
+{
+    pub fn new(range: Range<T>) -> Self {
+        RangeAllocator {
+            initial_range: range.clone(),
+            free_ranges: vec![range.clone()],
+        }
+    }
+
+    pub fn initial_range(&self) -> Range<T> {
+        self.initial_range.clone()
+    }
+
+    pub fn allocate_range(&mut self, length: T) -> Option<Range<T>> {
+        let mut best_fit: Option<(usize, Range<T>)> = None;
+        for range in self.free_ranges.iter().cloned().enumerate() {
+            let (index, range) = range;
+            let range_length = range.end - range.start;
+            if range_length < length {
+                continue;
+            } else if range_length == length {
+                // Found a perfect fit, so stop looking.
+                best_fit = Some((index, range));
+                break;
+            }
+            match best_fit {
```

Comment:
Can we do something like
```rust
best_fit = Some(match best_fit { .. })
```
here? might be a bit nicer

Reply:
No, the borrow checker disallows this.  If you try to fill in that match statement it gets convoluted pretty quickly.

Reply:
Actually disregard that, I figured it out.